src/lib/threadpool.c

```diff
@@ -0,0 +1,529 @@
+		abort();
+	uv_cond_destroy(&pi->planner_cond);
+	POST(empty(&pi->ordered) && empty(&pi->unordered));
+
+	for (i = 0; i < pi->threads_nr; i++) {
+		uv_cond_signal(&ts[i].cond);
+		if (uv_thread_join(&ts[i].thread))
+			abort();
+		POST(empty(&ts[i].inq));
+		uv_cond_destroy(&ts[i].cond);
+	}
+
+	free(pi->threads);
+	uv_mutex_destroy(&pi->mutex);
+	pi->threads_nr = 0;
+}
+
+static void pool_threads_init(pool_t *pool)
+{
+	uint32_t       i;
+	uv_sem_t       sem;
+	pool_impl_t   *pi = pool->pi;
+	pool_thread_t *ts;
+	struct targs   pa = {
+		.sem = &sem,
+		.pi = pi,
+	};
+	uv_thread_options_t config = {
+		.flags = UV_THREAD_HAS_STACK_SIZE,
+		.stack_size = 8u << 20,
+	};
+
+	if (uv_mutex_init(&pi->mutex))
+		abort();
+	if (uv_sem_init(&sem, 0))
+		abort();
+
+	pi->threads = calloc(pi->threads_nr, sizeof(pi->threads[0]));
+	if (pi->threads == NULL)
+		abort();
+
+	for (i = 0, ts = pi->threads; i < pi->threads_nr; i++) {
+		ts[i].arg = (struct targs) {
+			.pi = pi,
+			.sem = &sem,
+			.idx = i,
+		};
+
+		QUEUE__INIT(&ts[i].inq);
+		if (uv_cond_init(&ts[i].cond))
+			abort();
+		if (uv_thread_create_ex(&ts[i].thread, &config, worker,
+					&ts[i].arg))
+			abort();
+	}
+
+	if (uv_cond_init(&pi->planner_cond))
+		abort();
+	if (uv_thread_create_ex(&pi->planner_thread, &config, planner, &pa))
+		abort();
+	for (i = 0; i < pi->threads_nr + 1 /* +planner */; i++)
+		uv_sem_wait(&sem);
+
+	uv_sem_destroy(&sem);
+}
+
+static void pool_work_submit(pool_t *pool, pool_work_t *w)
+{
+	pool_impl_t *pi = pool->pi;
+	queue *o = &pi->ordered;
+	queue *u = &pi->unordered;
+
+	if (w->type > WT_UNORD) {
+		/* Make sure that elements in the ordered queue come in order. */
+		PRE(ERGO(pi->ord_prev != WT_BAR && w->type != WT_BAR,
+			 pi->ord_prev == w->type));
+		pi->ord_prev = w->type;
+	}
+
+	uv_mutex_lock(&pi->mutex);
+	push(w->type == WT_UNORD ? u : o, &w->link);
+	uv_cond_signal(&pi->planner_cond);
+	uv_mutex_unlock(&pi->mutex);
+}
+
+void work_done(uv_async_t *handle)
+{
+	queue q = {};
+	pool_impl_t *pi = container_of(handle, pool_impl_t, outq_async);
+
+	uv_mutex_lock(&pi->outq_mutex);
+	QUEUE__MOVE(&pi->outq, &q);
+	uv_mutex_unlock(&pi->outq_mutex);
+
+	while (!empty(&q))
+		queue_done(q_to_w(pop(&q)));
+}
+
+void pool_queue_work(pool_t *pool,
+		     pool_work_t *w,
+		     uint32_t cookie,
+		     enum pool_work_type type,
+		     void (*work_cb)(pool_work_t *w),
+		     void (*after_work_cb)(pool_work_t *w))
+{
+	PRE(work_cb != NULL && type < WT_NR);
+
+	*w = (pool_work_t) {
+		.pool = pool,
+		.type = type,
+		.thread_id = cookie % pool->pi->threads_nr,
+		.work_cb = work_cb,
+		.after_work_cb = after_work_cb,
+	};
+	w_register(pool, w);
+	pool_work_submit(pool, w);
+}
+
+int pool_init(pool_t *pool, uv_loop_t *loop, uint32_t threads_nr)
+{
+	int rc;
+	pool_impl_t *pi = pool->pi;
+
+	PRE(threads_nr <= THREADPOOL_SIZE_MAX);
+
+	pi = pool->pi = calloc(1, sizeof(*pool->pi));
+	if (pi == NULL)
+		return UV_ENOMEM;
+
+	*pi = (pool_impl_t) {
+		.qos = 0,
```

Comment:
How is this QoS functionality exposed to the user?

Reply:
OK, I'll expose it through pool_init().